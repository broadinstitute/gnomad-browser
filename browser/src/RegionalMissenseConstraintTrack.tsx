import React from 'react'
import styled from 'styled-components'

// @ts-expect-error TS(7016) FIXME: Could not find a declaration file for module '@gno... Remove this comment to see the full error message
import { Track } from '@gnomad/region-viewer'
import { TooltipAnchor } from '@gnomad/ui'

import Link from './Link'

import InfoButton from './help/InfoButton'
import { Gene } from './GenePage/GenePage'

type RegionalMissenseConstraintRegion = {
  chrom: string
  start: number
  stop: number
  region_start: number
  region_stop: number
  aa_start: string | null
  aa_stop: string | null
  obs_mis: number | undefined
  exp_mis: number
  obs_exp: number
  chisq_diff_null: number | undefined
  p_value: number
  z_score: number | undefined
}

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 1em;
`

const PlotWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 100%;
`

const RegionAttributeList = styled.dl`
  margin: 0;

  div {
    margin-bottom: 0.25em;
  }

  dt {
    display: inline;
    font-weight: bold;
  }

  dd {
    display: inline;
    margin-left: 0.5em;
  }
`

export const regionIntersections = (
  regionArrays: { start: number; stop: number }[][]
): RegionalMissenseConstraintRegion[] => {
  const sortedRegionsArrays = regionArrays.map((regions) =>
    [...regions].sort((a, b) => a.start - b.start)
  )

  const intersections = []

  const indices = sortedRegionsArrays.map(() => 0)

  while (sortedRegionsArrays.every((regions, i) => indices[i] < regions.length)) {
    const maxStart = Math.max(...sortedRegionsArrays.map((regions, i) => regions[indices[i]].start))
    const minStop = Math.min(...sortedRegionsArrays.map((regions, i) => regions[indices[i]].stop))

    if (maxStart < minStop) {
      const next = Object.assign(
        // @ts-ignore TS2556: A spread argument must either have a tuple type or be ...
        ...[
          {},
          ...sortedRegionsArrays.map((regions: { [x: string]: any }, i) => regions[indices[i]]),
          {
            start: maxStart,
            stop: minStop,
          },
        ]
      )

      intersections.push(next)
    }

    sortedRegionsArrays.forEach((regions, i) => {
      if (regions[indices[i]].stop === minStop) {
        indices[i] += 1
      }
    })
  }

  return intersections
}

// https://colorbrewer2.org/#type=sequential&scheme=YlOrRd&n=5
const colorScale = {
  not_significant: '#e2e2e2',
  least: '#9b001f',
  less: '#de351b',
  middle: '#fd8d3c',
  greater: '#fecc5c',
  greatest: '#ffffb2',
}

function regionColor(region: RegionalMissenseConstraintRegion) {
  if (region.z_score) {
    return region.z_score > 3.09 ? colorScale.middle : colorScale.not_significant
  }

  let color
  if (region.obs_exp > 0.8) {
    color = colorScale.greatest
  } else if (region.obs_exp > 0.6) {
    color = colorScale.greater
  } else if (region.obs_exp > 0.4) {
    color = colorScale.middle
  } else if (region.obs_exp > 0.2) {
    color = colorScale.less
  } else {
    color = colorScale.least
  }

  return region.p_value > 0.001 ? colorScale.not_significant : color
}

const LegendWrapper = styled.div`
  display: flex;

  @media (max-width: 600px) {
    flex-direction: column;
    align-items: center;
  }
`

const Legend = () => {
  return (
    <LegendWrapper>
      <span>Missense observed/expected</span>
      <svg width={170} height={25}>
        <rect x={10} y={0} width={30} height={10} stroke="#000" fill={colorScale.least} />
        <rect x={40} y={0} width={30} height={10} stroke="#000" fill={colorScale.less} />
        <rect x={70} y={0} width={30} height={10} stroke="#000" fill={colorScale.middle} />
        <rect x={100} y={0} width={30} height={10} stroke="#000" fill={colorScale.greater} />
        <rect x={130} y={0} width={30} height={10} stroke="#000" fill={colorScale.greatest} />
        <text x={10} y={10} fontSize="10" dy="1.2em" textAnchor="middle">
          0.0
        </text>
        <text x={40} y={10} fontSize="10" dy="1.2em" textAnchor="middle">
          0.2
        </text>
        <text x={70} y={10} fontSize="10" dy="1.2em" textAnchor="middle">
          0.4
        </text>
        <text x={100} y={10} fontSize="10" dy="1.2em" textAnchor="middle">
          0.6
        </text>
        <text x={130} y={10} fontSize="10" dy="1.2em" textAnchor="middle">
          0.8
        </text>
        <text x={160} y={10} fontSize="10" dy="1.2em" textAnchor="middle">
          1.0+
        </text>
      </svg>
      <svg width={170} height={25}>
        <rect x={10} y={0} width={20} height={10} stroke="#000" fill={colorScale.not_significant} />
        <text x={35} y={0} fontSize="10" dy="1em" textAnchor="start">
          Not significant (p &gt; 1e-3)
        </text>
      </svg>
    </LegendWrapper>
  )
}

const renderNumber = (number: number | undefined) => {
  return number === undefined || number === null ? '-' : number.toPrecision(4)
}

const renderNumberExponential = (number: number | undefined) => {
  return number === undefined || number === null ? '-' : number.toExponential(3)
}

const printAAorNA = (aa: string | null) => {
  if (aa === null) {
    return 'n/a'
  }
  return aa
}

type RegionTooltipProps = {
  region: RegionalMissenseConstraintRegion
  isTranscriptWide: boolean
}

const RegionTooltip = ({ region, isTranscriptWide }: RegionTooltipProps) => {
  if (isTranscriptWide) {
    return (
      <RegionAttributeList>
        <div>
          <dt>Missense observed/expected:</dt>
          <dd>{`${renderNumber(region.obs_exp)} (${region.obs_mis}/${renderNumber(
            region.exp_mis
          )})`}</dd>
        </div>
        <br />
        <div>The observed/expected ratio for this gene is transcript-wide.</div>
      </RegionAttributeList>
    )
  }
  return (
    <RegionAttributeList>
      <div>
        <dt>Coordinates:</dt>
        <dd>{`${region.chrom}:${region.region_start}-${region.region_stop}`}</dd>
      </div>
      <div>
        <dt>Amino acids:</dt>
        <dd>{`${printAAorNA(region.aa_start)}-${printAAorNA(region.aa_stop)}`}</dd>
      </div>
      <div>
        <dt>Missense observed/expected:</dt>
        <dd>{`${renderNumber(region.obs_exp)} (${region.obs_mis}/${renderNumber(
          region.exp_mis
        )})`}</dd>
      </div>
      <div>
        <dt>p-value:</dt>
        <dd>
          {renderNumberExponential(region.p_value)}
          {region.p_value !== null && region.p_value > 0.001 && ' (not significant)'}
        </dd>
      </div>
    </RegionAttributeList>
  )
}

const SidePanel = styled.div`
  display: flex;
  align-items: center;
  height: 100%;
`

const TopPanel = styled.div`
  display: flex;
  justify-content: flex-end;
  width: 100%;
  margin-bottom: 5px;
`

export type RegionalMissenseConstraint = {
  has_no_rmc_evidence: boolean
  passed_qc: boolean
  regions: RegionalMissenseConstraintRegion[]
}

type Props = {
  regionalMissenseConstraint?: RegionalMissenseConstraint
  gene: Gene
}

type TrackProps = {
  scalePosition: (input: number) => number
  width: number
}

const RegionalMissenseConstraintTrack = ({ regionalMissenseConstraint, gene }: Props) => {
  if (
    !regionalMissenseConstraint ||
    regionalMissenseConstraint.regions === null ||
    (regionalMissenseConstraint.passed_qc === false &&
      regionalMissenseConstraint.has_no_rmc_evidence === false)
  ) {
    return (
      <Track
        renderLeftPanel={() => (
          <SidePanel>
            <span>Regional missense constraint</span>
            <InfoButton topic="regional-constraint" />
          </SidePanel>
        )}
      >
        {({ width }: { width: number }) => (
          <>
            <PlotWrapper>
              <svg height={35} width={width}>
                <text x={width / 2} y={35 / 2} dy="1.0rem" textAnchor="middle">
                  <tspan>
                    This gene was not searched for evidence of regional missense constraint. See our{' '}
                  </tspan>
                  <tspan fill="#0000ff">
                    <Link to="/help">help page</Link>
                  </tspan>
                  <tspan> for additional information.</tspan>
                </text>
              </svg>
            </PlotWrapper>
          </>
        )}
      </Track>
    )
  }

  // This transcript was searched, but no RMC evidence was found
  //   instead, use the available gene level constraint data to display a single
  //   region for the RMC track
  if (regionalMissenseConstraint.has_no_rmc_evidence) {
    // eslint-disable-next-line no-param-reassign
    regionalMissenseConstraint.regions = []

    if (gene.gnomad_constraint) {
      // eslint-disable-next-line no-param-reassign
      regionalMissenseConstraint.regions = [
        {
          chrom: gene.chrom,
          start: Math.min(gene.start, gene.stop),
          stop: Math.max(gene.start, gene.stop),
          region_start: Math.min(gene.start, gene.stop),
          region_stop: Math.max(gene.start, gene.stop),
          obs_mis: gene.gnomad_constraint.obs_mis,
          exp_mis: gene.gnomad_constraint.exp_mis,
          obs_exp: gene.gnomad_constraint.oe_mis,
          z_score: gene.gnomad_constraint.mis_z,
          p_value: -0.01,
          chisq_diff_null: undefined,
          aa_start: null,
          aa_stop: null,
        },
      ]
    }
  }

  const constrainedExons = regionIntersections([
    regionalMissenseConstraint.regions.map((region) => {
      return {
        ...region,
        region_start: region.start,
        region_stop: region.stop,
      }
    }),
    gene.exons.filter((exon) => exon.feature_type === 'CDS'),
  ])

  return (
    <Wrapper>
      <Track
        renderLeftPanel={() => (
          <SidePanel>
            <span>Regional missense constraint</span>
            <InfoButton topic="regional-constraint" />
          </SidePanel>
        )}
      >
        {({ scalePosition, width }: TrackProps) => (
          <>
            <TopPanel>
              <Legend />
            </TopPanel>
            <PlotWrapper>
              <svg height={55} width={width}>
                {constrainedExons.map((region: RegionalMissenseConstraintRegion) => {
                  const startX = scalePosition(region.start)
                  const stopX = scalePosition(region.stop)
                  const regionWidth = stopX - startX

                  return (
                    <TooltipAnchor
                      key={`${region.start}-${region.stop}`}
                      // @ts-expect-error - from TooltipAnchor component of GBTK
                      region={region}
                      isTranscript={regionalMissenseConstraint.regions.length === 1}
                      tooltipComponent={RegionTooltip}
                    >
                      <g>
                        <rect
                          x={startX}
                          y={0}
                          width={regionWidth}
                          height={15}
                          fill={regionColor(region)}
                          stroke="black"
                        />
                      </g>
                    </TooltipAnchor>
                  )
                })}
                <g transform="translate(0,20)">
                  {regionalMissenseConstraint.regions.map(
                    (region: RegionalMissenseConstraintRegion, index: number) => {
                      const startX = scalePosition(region.start)
                      const stopX = scalePosition(region.stop)
                      const regionWidth = stopX - startX
                      const midX = (startX + stopX) / 2
                      // const offset = index * 15
                      const offset = index * 0

                      return (
                        <g key={`${region.start}-${region.stop}`}>
                          <line
                            x1={startX}
                            y1={2 + offset}
                            x2={startX}
                            y2={11 + offset}
                            stroke="#424242"
                          />
                          <line
                            x1={startX}
                            y1={7 + offset}
                            x2={stopX}
                            y2={7 + offset}
                            stroke="#424242"
                          />
                          <line
                            x1={stopX}
                            y1={2 + offset}
                            x2={stopX}
                            y2={11 + offset}
                            stroke="#424242"
                          />
                          {regionWidth > 40 && (
                            <>
                              <rect
                                x={midX - 15}
                                y={3 + offset}
                                width={30}
                                height={5}
                                fill="#fafafa"
                              />
                              <text x={midX} y={8 + offset} dy="0.33em" textAnchor="middle">
                                {region.obs_exp.toFixed(2)}
                              </text>
                            </>
                          )}
                        </g>
                      )
                    }
                  )}
                </g>
              </svg>
            </PlotWrapper>
          </>
        )}
      </Track>
    </Wrapper>
  )
}

RegionalMissenseConstraintTrack.defaultProps = {
  height: 15,
}

export default RegionalMissenseConstraintTrack
