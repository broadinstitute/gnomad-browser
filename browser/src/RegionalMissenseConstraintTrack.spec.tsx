import { describe, expect, test } from '@jest/globals'

// eslint-disable-next-line import/no-unresolved, import/extensions
import { regionIntersections } from './RegionalMissenseConstraintTrack'

describe('RegionaMissenseConstraintTrack', () => {
  test('regionIntersections', () => {
    const testCases = [
      {
        regions1: [
          { start: 2, stop: 4, i: 1 },
          { start: 6, stop: 8, i: 2 },
        ],
        regions2: [
          { start: 2, stop: 6, j: 1 },
          { start: 6, stop: 9, j: 2 },
        ],
        expected: [
          { start: 2, stop: 4, i: 1, j: 1 },
          { start: 6, stop: 8, i: 2, j: 2 },
        ],
      },
      {
        regions1: [
          { start: 2, stop: 4, i: 1 },
          { start: 6, stop: 8, i: 2 },
        ],
        regions2: [{ start: 1, stop: 8, j: 1 }],
        expected: [
          { start: 2, stop: 4, i: 1, j: 1 },
          { start: 6, stop: 8, i: 2, j: 1 },
        ],
      },
      {
        regions1: [
          { start: 2, stop: 4, i: 1 },
          { start: 6, stop: 8, i: 2 },
        ],
        regions2: [
          { start: 1, stop: 3, j: 1 },
          { start: 3, stop: 5, j: 2 },
          { start: 5, stop: 9, j: 3 },
        ],
        expected: [
          { start: 2, stop: 3, i: 1, j: 1 },
          { start: 3, stop: 4, i: 1, j: 2 },
          { start: 6, stop: 8, i: 2, j: 3 },
        ],
      },
      {
        regions1: [
          { start: 2, stop: 4, region_start: 2, region_stop: 4 },
          { start: 6, stop: 8, region_start: 6, region_stop: 8 },
        ],
        regions2: [
          { start: 1, stop: 3, j: 1 },
          { start: 3, stop: 5, j: 2 },
          { start: 5, stop: 9, j: 3 },
        ],
        expected: [
          { start: 2, stop: 3, region_start: 2, region_stop: 4, j: 1 },
          { start: 3, stop: 4, region_start: 2, region_stop: 4, j: 2 },
          { start: 6, stop: 8, region_start: 6, region_stop: 8, j: 3 },
        ],
      },
    ]

    testCases.forEach(({ regions1, regions2, expected }) => {
      expect(regionIntersections([regions1, regions2])).toEqual(expected)
    })
  })
})
