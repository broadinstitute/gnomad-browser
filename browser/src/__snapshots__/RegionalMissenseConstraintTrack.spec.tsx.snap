// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegionalMissenseConstraint has no unexpected changes when the RMC has evidence and passed QC 1`] = `
<div
  className="ConstraintTrack__Wrapper-sc-113yvfg-1 iBhwwK"
>
  <div
    className="Track__OuterWrapper-sc-1sdyh2h-0 bBMGlf"
  >
    <div
      className="Track__InnerWrapper-sc-1sdyh2h-1 cEOGGC"
    >
      <div
        className="Track__SidePanel-sc-1sdyh2h-3 iSYzDq"
        style={
          {
            "width": 100,
          }
        }
      >
        <div
          className="ConstraintTrack__SidePanel-sc-113yvfg-2 dJMFds"
        >
          <span>
            Regional missense constraint
          </span>
          <button
            className="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
            onClick={[Function]}
            type="button"
          >
            <img
              alt=""
              aria-hidden="true"
              src="test-file-stub"
            />
            <span
              style={
                {
                  "border": "0",
                  "clip": "rect(0 0 0 0)",
                  "height": "1px",
                  "margin": "-1px",
                  "overflow": "hidden",
                  "padding": "0",
                  "position": "absolute",
                  "whiteSpace": "nowrap",
                  "width": "1px",
                }
              }
            >
              More information
            </span>
          </button>
        </div>
      </div>
      <div
        className="Track__CenterPanel-sc-1sdyh2h-4 iAyrrk"
        style={
          {
            "width": 500,
          }
        }
      >
        <div
          className="ConstraintTrack__TopPanel-sc-113yvfg-3 imJcbb"
        >
          <div
            className="ConstraintTrack__LegendWrapper-sc-113yvfg-4 kkHtue"
          >
            <span>
              Missense observed/expected
            </span>
            <svg
              height={25}
              width={170}
            >
              <rect
                fill="#9b001f"
                height={10}
                stroke="#000"
                width={30}
                x={10}
                y={1}
              />
              <rect
                fill="#de351b"
                height={10}
                stroke="#000"
                width={30}
                x={40}
                y={1}
              />
              <rect
                fill="#fd8d3c"
                height={10}
                stroke="#000"
                width={30}
                x={70}
                y={1}
              />
              <rect
                fill="#fecc5c"
                height={10}
                stroke="#000"
                width={30}
                x={100}
                y={1}
              />
              <rect
                fill="#ffffb2"
                height={10}
                stroke="#000"
                width={30}
                x={130}
                y={1}
              />
              <text
                dy="1.2em"
                fontSize="10"
                textAnchor="middle"
                x={10}
                y={10}
              >
                0.0
              </text>
              <text
                dy="1.2em"
                fontSize="10"
                textAnchor="middle"
                x={40}
                y={10}
              >
                0.2
              </text>
              <text
                dy="1.2em"
                fontSize="10"
                textAnchor="middle"
                x={70}
                y={10}
              >
                0.4
              </text>
              <text
                dy="1.2em"
                fontSize="10"
                textAnchor="middle"
                x={100}
                y={10}
              >
                0.6
              </text>
              <text
                dy="1.2em"
                fontSize="10"
                textAnchor="middle"
                x={130}
                y={10}
              >
                0.8
              </text>
              <text
                dy="1.2em"
                fontSize="10"
                textAnchor="middle"
                x={160}
                y={10}
              >
                1.0+
              </text>
            </svg>
            <svg
              height={25}
              width={170}
            >
              <rect
                fill="#e2e2e2"
                height={10}
                stroke="#000"
                width={20}
                x={10}
                y={1}
              />
              <text
                dy="1em"
                fontSize="10"
                textAnchor="start"
                x={35}
                y={1}
              >
                Not significant (p &gt; 1e-3)
              </text>
            </svg>
          </div>
        </div>
        <div
          className="ConstraintTrack__PlotWrapper-sc-113yvfg-0 fHKmMU"
        >
          <svg
            height={55}
            width={500}
          >
            <g
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <rect
                fill="#e2e2e2"
                height={15}
                stroke="black"
                width={56}
                x={124}
                y={1}
              />
            </g>
            <g
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <rect
                fill="#e2e2e2"
                height={15}
                stroke="black"
                width={54}
                x={346}
                y={1}
              />
            </g>
            <g
              transform="translate(0,20)"
            >
              <g>
                <line
                  stroke="#424242"
                  x1={50}
                  x2={50}
                  y1={2}
                  y2={11}
                />
                <line
                  stroke="#424242"
                  x1={50}
                  x2={180}
                  y1={7}
                  y2={7}
                />
                <line
                  stroke="#424242"
                  x1={180}
                  x2={180}
                  y1={2}
                  y2={11}
                />
                <rect
                  fill="#fafafa"
                  height={5}
                  width={30}
                  x={100}
                  y={3}
                />
                <text
                  dy="0.33em"
                  textAnchor="middle"
                  x={115}
                  y={8}
                >
                  0.35
                </text>
              </g>
              <g>
                <line
                  stroke="#424242"
                  x1={250}
                  x2={250}
                  y1={2}
                  y2={11}
                />
                <line
                  stroke="#424242"
                  x1={250}
                  x2={400}
                  y1={7}
                  y2={7}
                />
                <line
                  stroke="#424242"
                  x1={400}
                  x2={400}
                  y1={2}
                  y2={11}
                />
                <rect
                  fill="#fafafa"
                  height={5}
                  width={30}
                  x={310}
                  y={3}
                />
                <text
                  dy="0.33em"
                  textAnchor="middle"
                  x={325}
                  y={8}
                >
                  0.37
                </text>
              </g>
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
